libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/UnitTestTools.hs

```diff
@@ -0,0 +1,184 @@
+  addr,
+  keyHashWitVKeyPairFor,
+  someKeyHashWitVKeyPair,
+  txOut,
+  rewardAccount,
+  emptyPState,
+  emptyDState,
+  emptyVState,
+  emptyCertState,
+  genLens,
+  arbitraryLens,
+  rewardCoin,
+  depositCoin,
+
+  -- * Types supported
+  Hash,
+  Addr (..),
+  RewardAccount (..),
+  Coin (..),
+  CompactForm (CompactCoin),
+  Credential (..),
+  StakeReference (..),
+  StandardCrypto,
+  Crypto,
+  DRep (DRepCredential),
+  KeyHash,
+  KeyRole (..),
+  SafeHash,
+  TxId,
+  TxIn,
+  IndividualPoolStake (IndividualPoolStake),
+  WitVKey,
+  RDPair (..),
+  UTxO (..),
+  PoolParams (..),
+  CertState (..),
+  DState (..),
+  PState (..),
+  VState (..),
+  StrictMaybe (..),
+) where
+
+import Cardano.Crypto.DSIGN.Class ()
+import Cardano.Ledger.Address (Addr (..), RewardAccount (..))
+import Cardano.Ledger.BaseTypes (Network (..), StrictMaybe (..))
+import Cardano.Ledger.CertState (CertState (..), DState (..), PState (..), VState (..))
+import Cardano.Ledger.Coin (Coin (..), CompactForm (CompactCoin))
+import Cardano.Ledger.Compactible (fromCompact)
+import Cardano.Ledger.Core
+import Cardano.Ledger.Credential (Credential (..), StakeReference (..))
+import Cardano.Ledger.Crypto (Crypto, StandardCrypto)
+import Cardano.Ledger.DRep (DRep (DRepCredential))
+import Cardano.Ledger.Keys (Hash, KeyHash (..), KeyRole (..), WitVKey (..), coerceKeyRole, hashKey)
+import Cardano.Ledger.PoolDistr (IndividualPoolStake (IndividualPoolStake))
+import Cardano.Ledger.PoolParams (PoolParams (..))
+import Cardano.Ledger.SafeHash (SafeHash)
+import Cardano.Ledger.TxIn (TxId (..), TxIn (..))
+import Cardano.Ledger.UMap (RDPair (..))
+import Cardano.Ledger.UTxO (UTxO (..))
+import Cardano.Ledger.Val (Val (..))
+import Data.Default.Class (Default (def))
+import Data.Kind (Type)
+import GHC.TypeNats (KnownNat, Nat)
+import Lens.Micro
+import Test.Cardano.Ledger.Core.Arbitrary ()
+import Test.Cardano.Ledger.Core.KeyPair (KeyPair (..), mkWitnessVKey)
+import Test.QuickCheck
+
+-- ==================================================
+
+-- | Supports patterns like
+--   (tom, john, ann, ron, mary) <- fresh @5 @(Credential 'Staking (EraCrypto era))
+class KnownNat n => Fresh (n :: Nat) (a :: Type) where
+  type Tuple n a :: Type
+  fresh :: Arbitrary a => Tuple n a
+
+instance Fresh 1 a where
+  type Tuple 1 a = Gen a
+  fresh = arbitrary @a
+
+instance Fresh 2 a where
+  type Tuple 2 a = Gen (a, a)
+  fresh = (arbitrary @(a, a))
+
+instance Fresh 3 a where
+  type Tuple 3 a = Gen (a, a, a)
+  fresh = (arbitrary @(a, a, a))
+
+instance Fresh 4 a where
+  type Tuple 4 a = Gen (a, a, a, a)
+  fresh = (arbitrary @(a, a, a, a))
+
+instance Fresh 5 a where
+  type Tuple 5 a = Gen (a, a, a, a, a)
+  fresh = (arbitrary @(a, a, a, a, a))
+
+instance Fresh 6 a where
+  type Tuple 6 a = Gen (a, a, a, a, a, a)
+  fresh = (arbitrary @(a, a, a, a, a, a))
+
+instance Fresh 7 a where
+  type Tuple 7 a = Gen (a, a, a, a, a, a, a)
+  fresh = (arbitrary @(a, a, a, a, a, a, a))
```

Comment:
This solution is way too complicated than it has to be. This is all it takes to do the same thing:
```haskell
type family TupleN (n :: Nat) a where
  TupleN 0 a = ()
  TupleN 1 a = a
  TupleN 2 a = (a, a)
  TupleN 3 a = (a, a, a)
  TupleN 4 a = (a, a, a, a)
  TupleN 5 a = (a, a, a, a, a)
  TupleN 6 a = (a, a, a, a, a, a)
  TupleN 7 a = (a, a, a, a, a, a, a)
```

Then you can use it in a very similar manner:
```haskell
myGen = do
  (x1, x2, x3) <- arbitrary @(TupleN 3 Int)
  pure $ x1 + x2 + x3
```

Note that this solution is not even specific to `QuickCheck`.

If you ok with switching to what I proposed then we can put `TupleN` into `cardano-data` package: `Data.TupleN`

Reply:
So I called the new package Data.MonoTuple